src/software/new_geom/util/distance.cpp

```diff
@@ -11,3 +11,91 @@ double distance(const Point &first, const Line &second)
 {
     return distance(second, first);
 }
+
+double distance(const Point &first, const Point &second)
+{
+    return (first - second).length();
+}
+
+double distance(const Segment &first, const Segment &second)
+{
+    if (intersects(first, second))
+    {
+        return 0.0;
+    }
+    return std::sqrt(std::min(std::min(distanceSquared(first, second.getSegStart()),
+                                       distanceSquared(first, second.getEnd())),
+                              std::min(distanceSquared(second, first.getSegStart()),
+                                       distanceSquared(second, first.getEnd()))));
```

Comment:
Can we split this out into a few variables to make it more readable?

Reply:
Done